### ISSUE TITLE
Change internode_compression default to 'dc'

### ISSUE THREAD

**Issue by tzach** [scylladb/scylladb#6712]
Traffic between DC is much more expensive and less sensitive to latency.
It make sense to enable it by default.

```
# internode_compression controls whether traffic between nodes is
# compressed.
# can be:  all  - all traffic is compressed
#          dc   - traffic between different datacenters is compressed
#          none - nothing is compressed.
# internode_compression: none
```

**Comment by nyh** [comment]
I can trivially change this default, but I need some advice from bigger experts what kind of implications this change can have:
1. Are we comfortable with the performance and stability of the compression option? (has anyone been using it in practice? testing it?)
2. Will rolling upgrade work? If a new node has compression enabled, but an old node sends it *uncompressed*  data, will the new node accept it? Will the old node accept compressed data even though it prefers uncompressed? I'll look into this myself, but maybe someone (@gleb-cloudius ?) already knows the answer off-hand.

**Comment by gleb-cloudius** [comment]
On Tue, Jun 29, 2021 at 06:02:44AM -0700, nyh wrote:
> I can trivially change this default, but I need some advice from bigger experts what kind of implications this change can have:
> 1. Are we comfortable with the performance and stability of the compression option? (has anyone been using it in practice? testing it?)
Its worthiness depends on cross DC latency. CPU utilization will be
larger as well.

> 2. Will rolling upgrade work? If a new node has compression enabled, but an old node sends it *uncompressed*  data, will the new node accept it? Will the old node accept compressed data even though it prefers uncompressed? I'll look into this myself, but maybe someone ***@***.*** ?) already knows the answer off-hand.
> 
IIRC the ability is negotiated (but we need to test obviously).

--
			Gleb.

**Comment by mykaul** [comment]
> internode_compression



> On Tue, Jun 29, 2021 at 06:02:44AM -0700, nyh wrote: I can trivially change this default, but I need some advice from bigger experts what kind of implications this change can have: 1. Are we comfortable with the performance and stability of the compression option? (has anyone been using it in practice? testing it?)
> Its worthiness depends on cross DC latency. CPU utilization will be larger as well.

I think we can safely assume there IS latency between DCs - and moreover, there's usually cost associated with that traffic, so there's more than one reason to enable it.
As for CPU utilization - if you send 20-30% (assuming) less packets and data, that helps a bit offset some of the CPU utilization, no?

> 2. Will rolling upgrade work? If a new node has compression enabled, but an old node sends it *uncompressed* data, will the new node accept it? Will the old node accept compressed data even though it prefers uncompressed? I'll look into this myself, but maybe someone ***@***.*** ?) already knows the answer off-hand.
> IIRC the ability is negotiated (but we need to test obviously).
> […](#)
> -- Gleb.

**Comment by mykaul** [comment]
@eliransin - let's revisit this for 5.3. Perhaps the cloud defaults should be different or perhaps the overall default should be changed.

**Comment by nyh** [comment]
I'm changing the tag of this issue. It's a performance issue, not "guardrails" (?) or "ux" (??) issue. It would have been guardrails if the request was to prevent the user from choosing uncompressed - but here we want to change the default.

**Comment by mykaul** [comment]
> I'm changing the tag of this issue. It's a performance issue, not "guardrails" (?) or "ux" (??) issue. It would have been guardrails if the request was to prevent the user from choosing uncompressed - but here we want to change the default.

Indeed. Maybe we should also add 'cost-optimization' label as well.

**Comment by mykaul** [comment]
> @mykaul I guess we need to throw it back to the backlog until it will get priority & commitment.

@DoronArazii - please open an issue to cloud - their default should be changed.

**Comment by bhalevy** [comment]
@mykaul are we still behind this issue?
Also, I opened https://github.com/scylladb/scylladb/issues/27085 to offer `rack` option, that could be useful for single dc clusters.

**Comment by michoecho** [comment]
> [@mykaul](https://github.com/mykaul) are we still behind this issue? Also, I opened [#27085](https://github.com/scylladb/scylladb/issues/27085) to offer `rack` option, that could be useful for single dc clusters.

@bhalevy Assuming AWS, `rack` is the only sensible option both for single-DC and multi-DC clusters.

Cross-DC (cross-region) traffic on AWS costs exactly as much as cross-rack (cross-AZ) traffic. And — barring weird setups with RF==1 or RF==0 DCs — there is always more cross-rack traffic than cross-DC traffic, because every time you send a write or read to another DC, it gets translated into at least two cross-rack replica calls within that DC.

**Comment by mykaul** [comment]
> [@mykaul](https://github.com/mykaul) are we still behind this issue? Also, I opened [#27085](https://github.com/scylladb/scylladb/issues/27085) to offer `rack` option, that could be useful for single dc clusters.

This is a dup of the same ask from me (https://github.com/scylladb/scylla-enterprise/issues/4316 ).

**Comment by michoecho** [comment]
> > [@mykaul](https://github.com/mykaul) are we still behind this issue? Also, I opened [#27085](https://github.com/scylladb/scylladb/issues/27085) to offer `rack` option, that could be useful for single dc clusters.
> 
> This is a dup of the same ask from me ([scylladb/scylla-enterprise#4316](https://github.com/scylladb/scylla-enterprise/issues/4316) ).

It first better in this repository, though.

**Comment by mykaul** [comment]
> > [@mykaul](https://github.com/mykaul) are we still behind this issue? Also, I opened [#27085](https://github.com/scylladb/scylladb/issues/27085) to offer `rack` option, that could be useful for single dc clusters.
> 
> [@bhalevy](https://github.com/bhalevy) Assuming AWS, `rack` is the only sensible option both for single-DC and multi-DC clusters.
> 
> Cross-DC (cross-region) traffic on AWS costs exactly as much as cross-rack (cross-AZ) traffic. And — barring weird setups with RF==1 or RF==0 DCs — there is always more cross-rack traffic than cross-DC traffic, because every time you send a write or read to another DC, it gets translated into at least two cross-rack replica calls within that DC.

I don't think that's correct: cross-region is $0.02 and cross-AZ is $0.01 IIRC. Ref. https://aws.amazon.com/ec2/pricing/on-demand/#Data_Transfer

**Comment by michoecho** [comment]
> I don't think that's correct: cross-region is $0.02 and cross-AZ is $0.01 IIRC. Ref. https://aws.amazon.com/ec2/pricing/on-demand/#Data_Transfer

@mykaul The catch is that cross-region traffic is charged $0.02 in only one direction ("Data Transfer OUT"), while cross-AZ traffic is charged $0.01 _in each direction_ ("Data transfer in", "Data transfer out"). So in both cases the cluster ends up paying $0.02 per GB.

**Comment by elcallio** [comment]
Is the consensus to do this or not?

**Comment by bhalevy** [comment]
> Is the consensus to do this or not?

Let's run a perfoemance benchmark with vs. without internode_compression=rack on AWS, with real (not simulated) AZs and then see.

**Comment by bhalevy** [comment]
Cc @mitso23

**Comment by mykaul** [comment]
The feature was implemented in:
Commit: 3c4546d839 (3c4546d8394528dd92c4e3b9363677d30f178e12)  
Author: Calle Wilund  
Date: Tue Nov 18 2025  
Message: messaging_service: Add internode_compression=rack as option
